service/src/main/java/io/camunda/service/ElementInstanceServices.java

```diff
@@ -25,35 +27,42 @@
 import java.util.Map;
 import java.util.concurrent.CompletableFuture;
 import java.util.function.Function;
+import java.util.stream.Collectors;
 
 public final class ElementInstanceServices
     extends SearchQueryService<
         ElementInstanceServices, FlowNodeInstanceQuery, FlowNodeInstanceEntity> {
 
   private final FlowNodeInstanceSearchClient flowNodeInstanceSearchClient;
+  private final ProcessCache processCache;
 
   public ElementInstanceServices(
       final BrokerClient brokerClient,
       final SecurityContextProvider securityContextProvider,
       final FlowNodeInstanceSearchClient flowNodeInstanceSearchClient,
+      final ProcessCache processCache,
       final CamundaAuthentication authentication) {
     super(brokerClient, securityContextProvider, authentication);
     this.flowNodeInstanceSearchClient = flowNodeInstanceSearchClient;
+    this.processCache = processCache;
   }
 
   @Override
   public ElementInstanceServices withAuthentication(final CamundaAuthentication authentication) {
     return new ElementInstanceServices(
-        brokerClient, securityContextProvider, flowNodeInstanceSearchClient, authentication);
+        brokerClient,
+        securityContextProvider,
+        flowNodeInstanceSearchClient,
+        processCache,
+        authentication);
   }
 
   @Override
   public SearchQueryResult<FlowNodeInstanceEntity> search(final FlowNodeInstanceQuery query) {
-    return flowNodeInstanceSearchClient
-        .withSecurityContext(
-            securityContextProvider.provideSecurityContext(
-                authentication, Authorization.of(a -> a.processDefinition().readProcessInstance())))
-        .searchFlowNodeInstances(query);
+    return search(
+        query,
+        securityContextProvider.provideSecurityContext(
+            authentication, Authorization.of(a -> a.processDefinition().readProcessInstance())));
```

Comment:
💭 We could even move this into the `search` method. The permission is used for both search and getByKey then. No permission => nothing found. For getByKey we would run into NOT_FOUND then.